SEMANTICS.md

```diff
@@ -0,0 +1,297 @@
+
+[`x/wasm`](https://github.com/CosmWasm/wasmd/tree/master/x/wasm) is a custom
+Cosmos SDK module, which processes certain messages and uses them to upload,
+instantiate, and execute smart contracts. In particular, it accepts a properly
+signed
+[`MsgExecuteContract`](https://github.com/CosmWasm/wasmd/blob/master/proto/cosmwasm/wasm/v1beta1/tx.proto#L76-L89),
+routes it to
+[`Keeper.Execute`](https://github.com/CosmWasm/wasmd/blob/master/x/wasm/keeper/keeper.go#L311-L355),
+which loads the proper smart contract and calls `execute` on it. Note that this
+method may either return a success (with data and events) or an error. In the
+case of an error here, it will revert the entire transaction in the block. This
+is the context we find ourselves in when our contract receives the `execute`
+call.
+
+### Basic Execution
+
+When we implement a contract, we provide the following entry point:
+
+```rust
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> { }
+```
+
+With `DepsMut`, this can read and write to the backing `Storage`, as well as use
+the `Api` to validate addresses, and `Query` the state of other contracts or
+native modules. Once it is done, it returns either `Ok(Response)` or
+`Err(ContractError)`. Let's examine what happens next:
+
+If it returns `Err`, this error is converted to a string representation
+(`err.to_string()`), and this is returned to the SDK module. _All state changes
+are reverted_ and `x/wasm` returns this error message, which will _generally_
+(see submessage exception below) abort the transaction, and return this same
+error message to the external caller.
+
+If it returns `Ok`, the `Response` object is parsed and processed. Let's look at
+the parts here:
+
+```rust
+pub struct Response<T = Empty>
+where
+    T: Clone + fmt::Debug + PartialEq + JsonSchema,
+{
+    /// Optional list of "subcalls" to make. These will be executed in order
+    /// (and this contract's subcall_response entry point invoked)
+    /// *before* any of the "fire and forget" messages get executed.
+    pub submessages: Vec<SubMsg<T>>,
+    /// After any submessages are processed, these are all dispatched in the host blockchain.
+    /// If they all succeed, then the transaction is committed. If any fail, then the transaction
+    /// and any local contract state changes are reverted.
+    pub messages: Vec<CosmosMsg<T>>,
+    /// The attributes that will be emitted as part of a "wasm" event
+    pub attributes: Vec<Attribute>,
+    pub data: Option<Binary>,
+}
+```
+
+In the Cosmos SDK, a transaction returns a number of events to the user, along
+with an optional data "result". This result is hashed into the next block hash
+to be provable and can return some essential state (although in general client
+apps rely on Events more). This result is more commonly used to pass results
+between contracts or modules in the sdk.
```

Comment:
Are events hashed into the app state as well?

Reply:
No they are not. There was a long discussion about that a few months ago.

Just `Data` and `Code` (0 = success) are hashed. They are stored in ResultHash (not AppHash, which is only the state the app writes, not other info that goes to tendermint). I can explain this a bit more.

Reply:
Alright, thanks. I don't think we need to go into details here but worth making that fact explicit.